### PR TITLE
Correctly handle buffer boundaries while decoding escape sequences from json stream

### LINKED ISSUE #1702: 1.3.0 randomly crashs  in decodeFromStream mode
```
2021-09-25 09:43:15.063 9938-9938/com.cisystems.ssmp.arm_vmb2 E/coroutineEH: Unexpected JSON token at offset 16249: Unexpected EOF during unicode escape
    JSON input: .....0432 \u041c\u041e","address":"\u0440-\u043d \u041a\u0440\u04.....
    kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 16249: Unexpected EOF during unicode escape
    JSON input: .....0432 \u041c\u041e","address":"\u0440-\u043d \u041a\u0440\u04.....
```

if i move to  decodeFromString then all is ok. My implement

```

  override fun responseBodyConverter(
        type: Type,
        annotations: Array<out Annotation>,
        retrofit: Retrofit,
    ): Converter<ResponseBody, *> =
        Converter { value: ResponseBody  ->
            J.decodeFromStream(J.serializersModule.serializer(type).nullable, value.byteStream() )
//            J.decodeFromString(J.serializersModule.serializer(type).nullable, value.bytes().decodeToString() )
        }


```

### PR DESCRIPTION
**Pull request by sandwwraith**
Fixes #1702

### CODE REVIEW

**Comment by slavonnet** [comment]
Have shapshot for testing? I can test it in alpha production env

**Comment by sandwwraith** [comment]
@slavonnet No, you'll have to build this particular branch by yourself (https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/building.md)

File: formats/json/commonMain/src/kotlinx/serialization/json/internal/lexer/AbstractJsonLexer.kt [hunk 1]
```diff
@@ -318,6 +318,9 @@ internal abstract class AbstractJsonLexer {
             if (char == STRING_ESC) {
                 usedAppend = true
                 currentPosition = appendEscape(lastPosition, currentPosition)
+                currentPosition = definitelyNotEof(currentPosition)
+                if (currentPosition == -1)
+                    fail("EOF", currentPosition)
```
[the remark this answers is not in the document]
Comment (qwwdfsad) [reply formats/json/commonMain/src/kotlinx/serialization/json/internal/lexer/AbstractJsonLexer.kt:323]: Also, these branches are almost identical, probably worth merging them as well

**Review by qwwdfsad — CHANGES_REQUESTED** [review]

File: formats/json/jvmTest/src/kotlinx/serialization/features/JsonJvmStreamsTest.kt [hunk 2]
```diff
@@ -65,4 +67,22 @@ class JsonJvmStreamsTest {
             Json.decodeViaStream(String.serializer(), "\"")
         }
     }
+
+    @Test
+    fun testRandomEscapeSequences()  {
+        repeat(10_000) {
```
Comment (qwwdfsad) [inline formats/json/jvmTest/src/kotlinx/serialization/features/JsonJvmStreamsTest.kt:73]: The test takes 30 secs on i9, probably 10k is a bit too much for a single unit-test

File: formats/json/commonMain/src/kotlinx/serialization/json/internal/lexer/AbstractJsonLexer.kt [hunk 3]
```diff
@@ -318,6 +318,9 @@ internal abstract class AbstractJsonLexer {
             if (char == STRING_ESC) {
                 usedAppend = true
                 currentPosition = appendEscape(lastPosition, currentPosition)
+                currentPosition = definitelyNotEof(currentPosition)
```
[the remark this answers is not in the document]
Comment (sandwwraith) [reply formats/json/commonMain/src/kotlinx/serialization/json/internal/lexer/AbstractJsonLexer.kt:321]: Good idea

File: formats/json/commonMain/src/kotlinx/serialization/json/internal/lexer/AbstractJsonLexer.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (sandwwraith) [reply formats/json/commonMain/src/kotlinx/serialization/json/internal/lexer/AbstractJsonLexer.kt:323]: I prefer not to merge them due to subtle `else if (++currentPosition >= source.length)` condition

[the remark this answers is not in the document]
Comment (sandwwraith) [reply formats/json/commonMain/src/kotlinx/serialization/json/internal/lexer/AbstractJsonLexer.kt:323]: `-1` won't be displayed (but the last X chars of input would be): https://github.com/Kotlin/kotlinx.serialization/blob/c0c60a633de9b1bc979308ca57d98715406f318d/formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonExceptions.kt#L24

https://github.com/Kotlin/kotlinx.serialization/blob/c0c60a633de9b1bc979308ca57d98715406f318d/formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonExceptions.kt#L32

File: formats/json/jvmTest/src/kotlinx/serialization/features/JsonJvmStreamsTest.kt (hunk 2 above)
Comment (sandwwraith) [reply formats/json/jvmTest/src/kotlinx/serialization/features/JsonJvmStreamsTest.kt:73]: I can reduce down to 1000, should be enough

**Review by qwwdfsad — APPROVED** [review]

**Pull request merged**